File: backend/faos_core/pipelines/approval.py
"""FAOS v6.0 — Approval & delivery gate (Python)."""

from __future__ import annotations

from typing import Any, Dict, Optional

from faos_core.connectors.webhooks import deliver_webhook


def format_client_facing_output(
    brief: Dict[str, Any], combined_text: str, qa: Dict[str, Any]
) -> str:
    return "\n".join(
        [
            f"# {brief.get('title')}",
            "",
            f"**Brand:** {brief.get('brand') or 'BulletsEye'}",
            f"**Category:** {brief.get('category')}",
            f"**QA Score:** {qa.get('score')}/100 "
            f"({'PASS' if qa.get('passed') else 'NEEDS REVIEW'})",
            "",
            "## Brief",
            str(brief.get("details") or ""),
            "",
            "## Deliverable",
            (combined_text or "").strip(),
            "",
            "---",
            "_Prepared by FAOS v6.0 Approval Gate_",
        ]
    )


async def run_approval_gate(
    *,
    pipeline_id: str,
    brief: Dict[str, Any],
    combined_text: str,
    qa: Dict[str, Any],
    auto_approve: Optional[bool] = None,
) -> Dict[str, Any]:
    client_facing = format_client_facing_output(brief, combined_text, qa)

    if not qa.get("passed"):
        return {
            "status": "pending",
            "client_facing_output": client_facing,
            "ready_for_publish": False,
        }

    auto = True if auto_approve is None else bool(auto_approve)
    status = "auto_approved" if auto else "approved"
    webhook = await deliver_webhook(
        {
            "event": "faos.pipeline.approved",
            "pipeline_id": pipeline_id,
            "brand": brief.get("brand"),
            "body": {
                "title": brief.get("title"),
                "category": brief.get("category"),
                "qa_score": qa.get("score"),
                "output": client_facing,
            },
        }
    )

    return {
        "status": status,
        "client_facing_output": client_facing,
        "ready_for_publish": True,
        "webhook": webhook,
    }

File: faos_core/pipelines/approval.ts
/**
 * FAOS v6.0 — Approval & delivery gate
 */

import { deliverWebhook } from "@/faos_core/connectors/webhooks";
import type { ClientBrief, QaScore } from "@/faos_core/types";

export type ApprovalPacket = {
  status: "pending" | "approved" | "rejected" | "auto_approved";
  client_facing_output: string;
  ready_for_publish: boolean;
  webhook?: Awaited<ReturnType<typeof deliverWebhook>>;
};

export function formatClientFacingOutput(
  brief: ClientBrief,
  combinedText: string,
  qa: QaScore
): string {
  return [
    `# ${brief.title}`,
    ``,
    `**Brand:** ${brief.brand || "BulletsEye"}`,
    `**Category:** ${brief.category}`,
    `**QA Score:** ${qa.score}/100 (${qa.passed ? "PASS" : "NEEDS REVIEW"})`,
    ``,
    `## Brief`,
    brief.details,
    ``,
    `## Deliverable`,
    combinedText.trim(),
    ``,
    `---`,
    `_Prepared by FAOS v6.0 Approval Gate_`,
  ].join("\n");
}

export async function runApprovalGate(opts: {
  pipeline_id: string;
  brief: ClientBrief;
  combinedText: string;
  qa: QaScore;
  auto_approve?: boolean;
}): Promise<ApprovalPacket> {
  const clientFacing = formatClientFacingOutput(
    opts.brief,
    opts.combinedText,
    opts.qa
  );

  if (!opts.qa.passed) {
    return {
      status: "pending",
      client_facing_output: clientFacing,
      ready_for_publish: false,
    };
  }

  const auto = opts.auto_approve !== false;
  const status = auto ? "auto_approved" : "approved";

  const webhook = await deliverWebhook({
    event: "faos.pipeline.approved",
    pipeline_id: opts.pipeline_id,
    brand: opts.brief.brand,
    body: {
      title: opts.brief.title,
      category: opts.brief.category,
      qa_score: opts.qa.score,
      output: clientFacing,
    },
  });

  return {
    status,
    client_facing_output: clientFacing,
    ready_for_publish: true,
    webhook,
  };
}
